src/main/java/com/purbon/search/fair/utils/AlphaAdjuster.java

```diff
@@ -0,0 +1,168 @@
+package com.purbon.search.fair.utils;
+
+import org.apache.commons.math3.distribution.BinomialDistribution;
+import java.util.ArrayList;
+
+public class AlphaAdjuster {
+
+
+    private int n;
+    private int k;
+    private double p;
+    private double alpha;
+    private int[] mTable;
+    private DataFrame auxMTable;
+    private int currentHigh;
+    private MTableGenerator mTableGenerator;
+
+    /*
+        Algorithm 1 of the k fair ranking Paper
+        @param k the size of the ranking to produce
+        @param p the expected proportion of protected elements
+        @param alpha the significance for each individual test
+        @return the probability of rejecting a fair ranking
+     */
+
+    public AlphaAdjuster(int n, int k, double p, double alpha) {
+        if (n < 1) {
+            throw new IllegalArgumentException("Parameter n must be at least 1");
+        }
+        if (p <= 0.0 || p >= 1.0) {
+            throw new IllegalArgumentException("Parameter p must be in ]0.0, 1.0[");
+        }
+        if (alpha <= 0.0 || alpha >= 1.0) {
+            throw new IllegalArgumentException("Parameter alpha must be in ]0.0, 1.0[");
+        }
+
+        this.n = n;
+        this.k = k;
+        this.p = p;
+        this.alpha = alpha;
+        this.currentHigh = 0;
+
+        this.mTableGenerator = new MTableGenerator(n, k, p, alpha);
+        this.mTable = this.mTableGenerator.getMTable();
+        this.auxMTable = this.computeAuxTMTable();
+    }
+
+    public DataFrame computeAuxTMTable() {
+        DataFrame table = new DataFrame("inv", "block");
+        int lastMSeen = 0;
+        int lastPosition = 0;
+        for (int position = 1; position < mTable.length; position++) {
+            if (mTable[position] == lastMSeen + 1) {
+                lastMSeen += 1;
+                table.put(position, position, (position - lastPosition));
+                lastPosition = position;
+            } else if (mTable[position] != lastMSeen) {
+                throw new RuntimeException("Inconsistent mtable");
+            }
+        }
+        table.resolveNullEntries();
+        return table;
+    }
+
+    public double computeSuccessProbability() {
+        int maxProtected = auxMTable.getLengthOf("inv") - 1;
+        int minProtected = 1;
+        double successProbability = 0;
+
+        ArrayList<Double> currentTrial;
+
+        double[] successObtainedProb = new double[maxProtected];
+        successObtainedProb = fillWithZeros(successObtainedProb);
+        successObtainedProb[0] = 1.0;
+        ArrayList<ArrayList<Double>> pmfCache = new ArrayList<>();
```

Comment:
pmfCache is the storage for the probability mass function for each trial. (each trial is a list of pmf values). I used arraylists only at the points where the size of the list is changing through the iterations. But i will take a closer look if this might be possible to use more arrays

Reply:
@tsuehr can you add a few comments here there describing the hole process, I am worried in a few months when we revisit this we are going to miss few important bits of the process.